File: app/services/audit.py
"""Audit log: who changed what, from where."""

from datetime import timedelta

from sqlalchemy.orm import Session

from app.db.base import session_scope, utcnow
from app.db.models import AuditEntry
from app.repos import misc
from app.services.access import AccessDenied, Principal

RETENTION = timedelta(days=365)


def log(
    s: Session,
    user_id: int | None,
    action: str,
    target: str = "",
    ip: str = "",
    **detail,
) -> None:
    """Record inside the caller's transaction; never store secrets in detail."""
    misc.audit(s, AuditEntry(user_id=user_id, action=action, target=target, ip=ip, detail=detail))


def entries(who: Principal) -> list[AuditEntry]:
    if not who.is_admin:
        raise AccessDenied("audit")

    with session_scope() as s:
        return misc.audit_page(s)


def prune() -> None:
    with session_scope() as s:
        misc.prune_audit(s, utcnow() - RETENTION)

File: internal/services/audit/audit.go
// Package audit records who changed what, from where.
package audit

import (
	"errors"
	"time"

	"andon/internal/db"
	"andon/internal/model"
	"andon/internal/repos/misc"
	"andon/internal/services/access"
)

// Retention is how long audit entries are kept before Prune removes them.
const Retention = 365 * 24 * time.Hour

// ErrDenied is returned by Entries for a non-admin principal.
var ErrDenied = errors.New("audit: admin only")

// Log records one action inside the caller's transaction. detail must never
// hold secrets: it is stored as-is and shown in the admin audit log.
func Log(q db.Queryer, userID *int64, action, target, ip string, detail map[string]any) error {
	return misc.Audit(q, &model.AuditEntry{
		At: time.Now().UTC(), UserID: userID, Action: action, Target: target, IP: ip, Detail: detail,
	})
}

// Entries returns the most recent page of audit entries. Admin only.
func Entries(q db.Queryer, who *access.Principal) ([]*model.AuditEntry, error) {
	if !who.IsAdmin() {
		return nil, ErrDenied
	}
	return misc.AuditPage(q, nil)
}

// Prune deletes audit entries older than Retention.
func Prune(q db.Queryer) error {
	return misc.PruneAudit(q, time.Now().UTC().Add(-Retention))
}
